app/src/services/attestation.ts

```diff
@@ -0,0 +1,430 @@
+          'No details provided.',
+          AttestationErrorCodes.FailedToConnectToAttestationAgent
+        )
+      }
+
+      const nonce = await this.fetchNonceForAttestation(connection)
+      if (!nonce) {
+        throw new BifoldError(
+          'Attestation Service',
+          'There was a problem with the attestation service.',
+          'No details provided.',
+          AttestationErrorCodes.FailedToFetchNonceForAttestation
+        )
+      }
+
+      const attestationObj = await this.generateAttestation(nonce)
+      if (!attestationObj) {
+        throw new BifoldError(
+          'Attestation Service',
+          'There was a problem with the attestation service.',
+          'No details provided.',
+          AttestationErrorCodes.FailedToGenerateAttestation
+        )
+      }
+
+      const result = this.requestAttestation(connection, attestationObj)
+
+      DeviceEventEmitter.emit(AttestationEventTypes.Completed, result)
+    } catch (error) {
+      this._attestationWorkflowInProgress = false
+      this.log?.error('Failed to fetch attestation credential', error as Error)
+
+      DeviceEventEmitter.emit(AttestationEventTypes.FailedRequestCredential, error)
+    }
+  }
+
+  private handleCredentialOfferReceived = async (event: BaseEvent) => {
+    const { credentialRecord } = event.payload
+    const credential = credentialRecord as CredentialExchangeRecord
+
+    try {
+      const { offer } = await this.agent.credentials.getFormatData(credential.id)
+      const offerData = (offer?.anoncreds ?? offer?.indy) as AnonCredsCredentialOffer
+
+      // do nothing if not an attestation credential
+      const { attestationCredDefIds } = this.options
+      if (!attestationCredDefIds.includes(offerData?.cred_def_id ?? '')) {
+        return
+      }
+
+      // if it's a new offer, automatically accept
+      if (credential.state === CredentialState.OfferReceived) {
+        this.log?.info('Accepting credential offer')
+        await this.agent.credentials.acceptOffer({
+          credentialRecordId: credential.id,
+        })
+      }
+
+      // only finish loading state once credential is fully accepted
+      if (credential.state === CredentialState.Done) {
+        this.log?.info('Credential accepted')
+      }
+    } catch (error) {
+      this.log?.error('Failed to handle credential offer', error as Error)
+
+      DeviceEventEmitter.emit(AttestationEventTypes.FailedHandleOffer, error)
+    }
+  }
+
+  private handleProofReceived = async (event: BaseEvent) => {
+    const { proofRecord } = event.payload
+    const proof = proofRecord as ProofExchangeRecord
+
+    if (proof.state !== ProofState.RequestReceived) {
+      return
+    }
+
+    this.log?.info('Checking if proof is requesting attestation')
+
+    try {
+      // 1. Is the proof requesting an attestation credential
+      if (!(await isProofRequestingAttestation(proof, this.agent, this.options.attestationCredDefIds))) {
```

Comment:
Is this check necessary if it is repeated within `this.attestationCredentialRequired` on line 408?

Reply:
They're a little different no? One is checking if a recently arrived proof request is requesting attestation. If not it can safely be ignored. The second (408) is checking to see if we already have an attestation credential or not? If we have one we don't need to start the process of getting one - we can just use it.

Reply:
I think I worded that poorly. Within the `attestationCredentialRequired` method, on line 432, there's a second call of `isProofRequestingAttestation` that seems to also check if the same proof is requesting attestation, and also does an early return if it's not. I could be confused though. Not a big deal if it gets called an extra time anyway, it takes less than 50ms.

Reply:
I removed the second check.